clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,914 @@
+
+            // If the error code indicates that the interval ms needs to be updated as per the error
+            // code then update the interval ms and state so that the subscription can be re-fetched,
+            // and the push retried.
+            if (errorIntervalMsOpt.isPresent()) {
+                try {
+                    lock.writeLock().lock();
+                    // This is the case when client began termination sometime after the last push request
+                    // was issued. Just getting the callback, hence need to ignore it.
+                    if (isTerminatingState()) {
+                        return;
+                    }
+
+                    if (!maybeSetState(ClientTelemetryState.SUBSCRIPTION_NEEDED)) {
+                        log.error("Unable to transition state after failed push telemetry from state {}", state);
+                    }
+                    updateErrorResult(errorIntervalMsOpt.get(), now);
+                    return;
+                } finally {
+                    lock.writeLock().unlock();
+                }
+            }
+
+            try {
+                lock.writeLock().lock();
+                lastRequestMs = now;
+                intervalMs = subscription.pushIntervalMs();
+                if (!maybeSetState(ClientTelemetryState.PUSH_NEEDED)) {
+                    log.error("Unable to transition state after successful push telemetry from state {}", state);
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+        }
+
+        @Override
+        public void handleFailedGetTelemetrySubscriptionsRequest(KafkaException maybeFatalException) {
+            log.warn("The broker generated an error for the get telemetry network API request", maybeFatalException);
+            handleFailedRequest(maybeFatalException != null);
+        }
+
+        @Override
+        public void handleFailedPushTelemetryRequest(KafkaException maybeFatalException) {
+            log.warn("The broker generated an error for the push telemetry network API request", maybeFatalException);
+            handleFailedRequest(maybeFatalException != null);
+        }
+
+        @Override
+        public Optional<Uuid> clientInstanceId(Duration timeout) {
+            long timeoutMs = timeout.toMillis();
+            if (timeoutMs < 0) {
+                throw new IllegalArgumentException("The timeout cannot be negative for fetching client instance id.");
+            }
+
+            try {
+                lock.writeLock().lock();
+                if (subscription == null) {
+                    // If we have a non-negative timeout and no-subscription, let's wait for one to be retrieved.
+                    log.debug("Waiting for telemetry subscription containing the client instance ID with timeoutMillis = {} ms.", timeoutMs);
+                    try {
+                        if (!subscriptionLoaded.await(timeoutMs, TimeUnit.MILLISECONDS)) {
```

Comment:
Isn't the await here potentially a deadlock?

Reply:
I may be missing something but doesn't seems like, I have added a test case as well to validate if we can get the clientInstanceId when concurrent calls are made.